pyfarm/jobtypes/core/jobtype.py

```diff
@@ -890,6 +890,96 @@ def set_task_progress(self, task, progress):
+
+        if "id" not in task or not isinstance(task["id"], INTEGER_TYPES):
+            raise TypeError(
+                "Expected to find 'id' in `task` or for `task['id']` to "
+                "be an integer.")
+
+        url = "%s/jobs/%s/tasks/%s" % (
+            config["master_api"], self.assignment["job"]["id"], task["id"])
+        data = {"time_started": "now"}
+
+        updated = False
+        num_retry_errors = 0
+        while not updated:
+            try:
+                response = yield post_direct(url, data=data)
+                response_data = yield treq.json_content(response)
+                if response.code == OK:
+                    logger.info("Set time_started of task %s to now on master",
+                                task ["id"])
+                    updated = True
+                    returnValue(None)
+
+                elif response.code >= INTERNAL_SERVER_ERROR:
+                    delay = http_retry_delay()
+                    logger.warning(
+                        "Could not post start time for task %s to the "
+                        "master server, internal server error: %s.  Retrying "
+                        "in %s seconds.", task["id"], response_data, delay)
+
+                    deferred = Deferred()
+                    reactor.callLater(delay, deferred.callback, None)
+                    yield deferred
+
+                elif response.code == NOT_FOUND:
+                    message = ("Got 404 NOT FOUND error on setting start time "
+                               "for task %s" % task["id"])
+                    logger.error(message)
+                    raise Exception(message)
+
+                elif response.code >= BAD_REQUEST:
+                    message = (
+                        "Failed to set start time for task %s on the "
+                        "master, bad request: %s. Server.  This request will "
+                        "not be retried." % (task["id"], response_data))
+                    logger.error(message)
+                    raise Exception(message)
+
+                else:
+                    message = (
+                        "Unhandled error when setting start time for task "
+                        "%s to the master: %s (code: %s).  This request will "
+                        "not be retried." % (response_data, response.code))
+                    logger.error(message)
+                    raise Exception(message)
+
+            except (ResponseNeverReceived, RequestTransmissionFailed) as error:
+                num_retry_errors += 1
+                if num_retry_errors > config["broken_connection_max_retry"]:
+                    message = (
+                        "Failed to set start time for task %s to the "
+                        "master, caught try-again type errors %s times in a "
+                        "row." % (task["id"], num_retry_errors))
+                    logger.error(message)
+                    raise Exception(message)
```

Comment:
Perhaps we should make this a custom exception, similar to what we did with `InsufficientDiskSpace`, so it can be more explicitly handled?